### PR TITLE
feat(hardfork): net hardfork

### PR DESCRIPTION
**Pull request by driftluo**
Modifications on the network are relatively simple but are closely related to the timing of the VM switch

1. All protocols support both versions 1 and 2, and all protocols are simple upgrades except for relay.
2. Relay protocols for runtime switching will have two mutually exclusive versions, and only one of them will work in conjunction whilst the relay Tx is in operation.
3. During switching, a loss of transactions is possible, which is almost an unavoidable problem.
4. Transactions that are executed in different versions of the VM may have inconsistent cycle consumption. To avoid this split network behaviour, some complex stuff needs to be done on the relay protocol
5. These unfriendly-looking network processing codes will, fortunately, be removed in the next release once the Fork feature is completed. Previous clients not supporting the Fork feature will be unable to connect to the network anymore, yet the first Fork-enabled client will be compatible with both networks for some time after the Fork is completed and will reject connections from older clients.

Test:
1. Communicates normally with old clients before the fork
2. Rejects connections from old clients after the fork
3. Evicts connections from old clients during the fork
4. Client after the fork allows nodes that support the fork feature to synchronize from scratch
5. No banning for inconsistent relay tx during the switching process

### CODE REVIEW

File: util/types/src/core/hardfork.rs [hunk 1]
```diff
@@ -209,6 +216,14 @@ define_methods!(
     "RFC PR 0237"
 );
 
+define_methods!(
+    rfc_pr_0234,
+    p2p_network_switch,
+    is_p2p_network_switch_enabled,
```
Comment (yangby-cryptape) [inline util/types/src/core/hardfork.rs:222]: I think we could use the same switch, since they have to equal to each other.

Comment (keroro520) [reply util/types/src/core/hardfork.rs:222]: I agree. https://github.com/nervosnetwork/ckb/pull/2765#discussion_r661295725

File: sync/src/types/mod.rs [hunk 2]
```diff
@@ -249,11 +249,24 @@ impl HeadersSyncController {
     }
 }
 
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub enum Version {
+    Old,
+    New,
+}
```
Comment (keroro520) [inline sync/src/types/mod.rs:256]: > A public type called Version is easy to misunderstand, protocol version, implementation version, binary version, or else?

https://github.com/nervosnetwork/ckb/pull/2765#discussion_r660292829

File: util/types/src/core/hardfork.rs (hunk 1 above)
Comment (driftluo) [reply util/types/src/core/hardfork.rs:222]: Change 00237 to 234 + 237, so it will share the same value

File: sync/src/types/mod.rs (hunk 2 above)
Comment (driftluo) [reply sync/src/types/mod.rs:256]: change to binary version

**Review by keroro520 — APPROVED** [review]
LGTM, expect the discovery protocol. I know it will change later.

File: sync/src/types/mod.rs [hunk 3]
```diff
@@ -249,11 +249,24 @@ impl HeadersSyncController {
     }
 }
 
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub enum BinaryVersion {
```
Comment (quake) [inline sync/src/types/mod.rs:253]: I don't think we need a new enum here, just add a new field to struct `PeerFlags`, which can reduce a large number of fn params changes:
```
pub struct PeerFlags {
    pub is_outbound: bool,
    pub is_protect: bool,
    pub is_whitelist: bool,
    pub is_2021edition: bool,
}
```

**Review by quake — CHANGES_REQUESTED** [review]

File: sync/src/types/mod.rs (hunk 3 above)
Comment (driftluo) [reply sync/src/types/mod.rs]: Modification is complete, please review again

**Review by quake — DISMISSED** [review]

**Review by doitian — APPROVED** [review]
Lgtm. Except one suggestion: avoid too many `v2: bool`, use an enum instead.

**Review by quake — APPROVED** [review]

**Comment by driftluo** [comment]
bors r=quake,doitian

**Pull request merged**